rye/src/cli/publish.rs

```diff
@@ -181,30 +194,161 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
     Ok(())
 }
 
-fn prompt_for_token() -> Result<String, Error> {
+struct Credentials {
+    username: Option<String>,
+    password: Option<Secret<String>>,
+}
+
+struct Repository {
+    name: Option<String>,
+    url: Option<Url>,
+}
+
+fn is_unknown_repository(repository: &Repository) -> bool {
+    repository.name.is_none() && repository.url.is_none()
+}
+
+fn is_default_repository(repository: &Repository) -> bool {
+    repository
+        .name
+        .as_ref()
+        .map_or(false, |it| it == DEFAULT_REPOSITORY)
+        || repository
+            .url
+            .as_ref()
+            .map_or(false, |it| it.domain() == Some(DEFAULT_REPOSITORY_DOMAIN))
+}
+
+fn is_keyring_ready(credentials: &Credentials, repository: &Repository) -> bool {
+    credentials.username.is_some() && repository.url.is_some()
+}
+
+fn resolve_credentials(
+    credentials_table: Option<&Item>,
+    username: Option<&String>,
+    password: Option<&Secret<String>>,
+) -> Credentials {
+    let mut credentials = Credentials {
+        username: None,
+        password: None,
+    };
+
+    if username.is_some() {
+        credentials.username = username.cloned();
+    } else {
+        credentials.username = credentials_table
+            .and_then(|it| it.get("username").map(Item::to_string).map(escape_string));
+    }
+
+    if password.is_some() {
+        credentials.password = password.cloned();
+    } else {
+        credentials.password = credentials_table.and_then(|it| {
+            it.get("token")
+                .map(Item::to_string)
+                .map(escape_string)
+                .map(Secret::new)
+        });
+    }
+
+    if credentials.username.is_some() && credentials.password.is_some() {
+        return credentials;
+    }
+
+    // Rye resolves tokens from the file or the cli. If a token was resolved
+    // we can assume a default username of __token__.
+    if credentials.password.is_some() && credentials.username.is_none() {
+        credentials.username = Some(DEFAULT_USERNAME.to_string())
+    }
+
+    credentials
+}
+
+fn resolve_repository(
+    credentials_table: Option<&Item>,
+    name: Option<String>,
+    url: Option<Url>,
+) -> Result<Repository, Error> {
+    let mut repository = Repository { name, url };
+
+    if repository.url.is_some() {
+        return Ok(repository);
+    }
+
+    if let Some(cred_url) = credentials_table.and_then(|it| {
+        it.get("repository-url")
+            .map(Item::to_string)
+            .map(escape_string)
+    }) {
+        repository.url = Some(Url::parse(&cred_url)?);
+    }
+
+    if repository.url.is_none()
+        && repository
+            .name
+            .as_ref()
+            .map_or(false, |it| it == DEFAULT_REPOSITORY)
+    {
+        repository.url = Some(Url::parse(DEFAULT_REPOSITORY_URL)?);
+    }
```

Comment:
`resolve_repository` will resolve a default url but it doesn't resolve a default repository name if the default url is found. This would be useful for writing mentioned [here](https://github.com/astral-sh/rye/pull/759/files#r1509867782).

Reply:
I've moved this into `Repository`'s `resolve_with_defaults`.